refactor(language-service): Cleanup TypescriptHost

kyliau opened the pull request:
Cleanup the logic in TypeScriptHost as to when langauge service state
should be synchronized with the editor state.

The model employed follows that of tsserver, in which case it is the
caller's responsiblity to synchronize host data before any LS methods
are called.

As part of this refactoring the integration test for external template is removed because we're no longer using tsserver plugin to handle HTML files.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

Commit:
refactor(language-service): Cleanup TypescriptHost

Cleanup the logic in TypeScriptHost as to when langauge service state
should be synchronized with the editor state.

The model employed follows that of tsserver, in which case it is the
caller's responsiblity to synchronize host data before any LS methods
are called.

andrius-pra commented on packages/language-service/src/reflector_host.ts:
```diff
@@ -63,14 +61,26 @@ export class ReflectorHost implements StaticSymbolResolverHost {
+        // the module without the caller explicitly providing a containing file.
+        throw new Error(`Could not resolve '${moduleName}' without a containing file.`);
+      }
+      containingFile = currentDirectory;
```
"ts.resolveModuleName" requires full filename in a 2nd parameter. This method returns wrong result when it get directory instead of path to file.

kyliau replied:
I double checked https://github.com/microsoft/TypeScript/blob/b70f89488193f403fcdb8e811bdaf690636658ff/src/compiler/moduleNameResolver.ts#L634

It looks like in both cases of `ModuleResolutionKind.NodeJs` and `ModuleResolutionKind.Classic`, only the directory path is used.

kyliau replied:
ah, looking at the code more closely, you're right. It does remove the "leaf" segment of the directory path regardless of file or directory.

andrius-pra commented on packages/language-service/src/language_service.ts:
```diff
@@ -32,9 +32,13 @@ class LanguageServiceImpl implements LanguageService {
 
-  getTemplateReferences(): string[] { return this.host.getTemplateReferences(); }
+  getTemplateReferences(): string[] {
+    this.host.getAnalyzedModules();  // same role as 'synchronizeHostData'
```
you also need `host.getAnalyzedModules();` [in getExternalFiles method (ts_plugin.ts)](https://github.com/kyliau/angular/blob/165c43c0b665894fa820d1989c996778e9d0f5d1/packages/language-service/src/ts_plugin.ts#L22)

kyliau replied:
Good call. Thank you!

Commit:
test(language-service): Remove test for external template

The tsserver is not meant to handle HTML files, so there is no point
sending an "open" request. The existing test is wrong because the
quickinfo returns "const name: never", which should be
"(property) WidgetComponent.name"

kyliau commented:
Caretaker: this PR was previously reviewed and approved by @ayazhafiz but he was removed from the Angular org. (Working with Misko to reinstate membership).
